sqlglot/optimizer/qualify_columns.py

```diff
@@ -495,6 +495,38 @@ def quote_identifiers(expression: E, dialect: DialectType = None, identify: bool
     )
 
 
+def pushdown_cte_alias_columns(expression: exp.Expression) -> exp.Expression:
+    """
+    Pushes down the CTE alias columns into the projection,
+
+    This step is useful in Snowflake where the CTE alias columns can be referenced in the HAVING.
+
+    Example:
+        >>> import sqlglot
+        >>> expression = sqlglot.parse_one("WITH y (c) AS (SELECT SUM(a) FROM ( SELECT 1 a ) AS x HAVING c > 0) SELECT c FROM y")
+        >>> pushdown_cte_alias_columns(expression).sql()
+        'WITH y(c) AS (SELECT SUM(a) AS c FROM (SELECT 1 AS a) AS x HAVING c > 0) SELECT c FROM y'
+
+    Args:
+        expression: Expression to pushdown.
+
+    Returns:
+        The expression with the CTE aliases pushed down into the projection.
+    """
+    for cte in expression.find_all(exp.CTE):
+        if cte.alias_column_names:
+            new_expressions = []
+            for _alias, projection in zip(cte.alias_column_names, cte.this.expressions):
```

Comment:
Might need special handling for UNIONs:

```sql
WITH x (c) AS (
    SELECT
      SUM(a)
    FROM x
    GROUP BY b
    HAVING c > 0

    UNION ALL

    SELECT
      SUM(a)
    FROM y
    GROUP BY b
    HAVING c > 0
) ...
```

I'm actually curious what Snowflake does here...

Also, I think it's possible to have subqueries in a CTE. Weird stuff like:
```sql
WITH x AS (
  (
    SELECT 1 a
  )
  LIMIT 1
)
SELECT * FROM x
```

And again, I'm curious what snowflake does in these cases.

Reply:
I tried the first case (without the GROUP BY b) after your initial comment and Snowflake throws an error so I didn't code for it. I tested with the following and it worked as expected:

```
WITH y (c) AS (
  SELECT sum(a) from (SELECT SUM(a) AS a
                      FROM (SELECT 1 a) AS x
                      UNION ALL
                      SELECT sum(d) as a
                      FROM (SELECT 1 d) as y) t
                                      HAVING c > 0
)
SELECT * FROM y;
```

The second case with the subqueries is valid but also not sure how the pushdown should work in this case. I tried turning into the WITH x(c) ... and adding the HAVING clause but that's not valid in Snowflake:

```
-- INVALID
WITH x(c) AS (
  (
    SELECT 1 a
  )
  HAVING c > 0
)
SELECT * FROM x;

-- VALID and handles the pushdown
WITH x(c) AS (
    SELECT sum(1) a
  HAVING c > 0
  LIMIT 1
)
SELECT * FROM x;"""

-- INVALID
WITH x(c) AS (
    (SELECT sum(1) a)
  HAVING c > 0
  LIMIT 1
)
SELECT * FROM x;
```

Reply:
Got it.
Let's add some tests for some of these in `tests/fixtures/optimizer/pushdown_cte_alias_columns.sql`, even if its a noop, to make sure it doesn't throw an error. I'm worried because this code makes an implicit assumption that its dealing with an `exp.Select` instance.